src/backend/access/transam/slru.c

```diff
@@ -774,22 +779,25 @@ SimpleLruDownloadSegment(SlruCtl ctl, int pageno, char const* path)
 			pfree(buffer);
 			return -1;
 		}
-		errno = 0;
-		pgstat_report_wait_start(WAIT_EVENT_SLRU_WRITE);
-		if (pg_pwrite(fd, buffer, n_blocks*BLCKSZ, 0) != n_blocks*BLCKSZ)
-		{
-			pgstat_report_wait_end();
-			/* if write didn't set errno, assume problem is no disk space */
-			if (errno == 0)
-				errno = ENOSPC;
-			slru_errcause = SLRU_WRITE_FAILED;
-			slru_errno = errno;
 
-			CloseTransientFile(fd);
-			pfree(buffer);
-			return -1;
+		if (!neon_use_communicator_worker) {
+			errno = 0;
+			pgstat_report_wait_start(WAIT_EVENT_SLRU_WRITE);
+			if (pg_pwrite(fd, buffer, n_blocks*BLCKSZ, 0) != n_blocks*BLCKSZ)
+			{
+				pgstat_report_wait_end();
+				/* if write didn't set errno, assume problem is no disk space */
+				if (errno == 0)
+					errno = ENOSPC;
+				slru_errcause = SLRU_WRITE_FAILED;
+				slru_errno = errno;
+
+				CloseTransientFile(fd);
+				pfree(buffer);
+				return -1;
+			}
+			pgstat_report_wait_end();
 		}
-		pgstat_report_wait_end();
 	}
 	pfree(buffer);
```

Comment:
while `pfree(NULL)` is not an error condition, it'd be nice to make cleanup conditional on us actually having a buffer to clean up.

Reply:
I was thinking about that as well, but since freeing a NULL pointer is perfectly valid in C, I thought it was fine.